padl/transforms.py

```diff
@@ -658,38 +663,71 @@ def _pd_unpack_argument(self, arg) -> bool:
             return self._pd_number_of_inputs > 1
 
         try:
-            parameters = self._pd_get_signature().values()
+            parameters = self._pd_get_signature.values()
         except ValueError:
             return False
         for param in parameters:
-            if param.kind in (
+            param_kind = param.kind
+            if param_kind in (
                     param.POSITIONAL_OR_KEYWORD,
                     param.POSITIONAL_ONLY):
                 signature_count += 1
-            if param.kind == param.VAR_POSITIONAL:
+            if param_kind == param.VAR_POSITIONAL:
                 return True
         if signature_count > 1:
             return True
         return False
 
-    def pd_call_transform(self, arg, mode: Optional[Mode] = None):
+    def _pd_unpack_args_and_call(self, arg):
+        if self._pd_unpack_argument(arg):
+            return self(*arg)
+        return self(arg)
+
+    def pd_call_in_mode(self, arg, mode: Mode, ignore_grad=False):
         """Call the transform, with possibility to pass multiple arguments.
 
         :param arg: argument to call the transform with
         :param mode: The mode ("infer", "eval", "train") to perform the call with.
+        :param ignore_grad: If *True* gradient settings are ignored
         :return: Whatever the transform returns.
         """
+        no_grad = mode in ('eval', 'infer') and not ignore_grad
 
-        if mode in ('eval', 'infer'):
-            torch_context = torch.no_grad()
-        else:
-            torch_context = contextlib.suppress()
+        layers = self.pd_layers
+        if not layers:
+            no_grad = False
+
+        if mode is not None:
+            Transform.pd_mode = mode
+            if layers:
+                training_before = [layer.training for layer in layers]
+                for layer in layers:
+                    if mode == 'train':
+                        layer.train()
+                    else:
+                        layer.eval()
+
+        if no_grad:
+            grad_before = torch.is_grad_enabled()
+            torch.set_grad_enabled(False)
 
-        with self.pd_set_mode(mode), torch_context:
-            if self._pd_unpack_argument(arg):
-                return self(*arg)
-            return self(arg)
+        try:
+            return self._pd_unpack_args_and_call(arg)
+        finally:
+            if mode is not None:
+                Transform.pd_mode = None
+                if layers:
+                    for i, training in enumerate(training_before):
+                        layer = layers[i]
+                        if training:
+                            layer.train()
+                        else:
+                            layer.eval()
+            if no_grad:
+                torch.set_grad_enabled(grad_before)
 
+    @property
+    @lru_cache(maxsize=128)
```

Comment:
@wuhu I changed this to be a property and cached since I believe it only needs to be evaluated once. Do you think this is correct?

Reply:
Agree to having the cache. If we make it a property we should rename it (to `_pd_signature`).

Reply:
Sounds good, I'll change this